File: 343-integer-break/343-integer-break.py
class Solution:
    def integerBreak(self, n: int) -> int:
        def f(n:int,dp):
            if n==1:
                return 1
            # elif n==2:
            #     return 1
            elif dp[n]!=-1:
                return dp[n]
            else:
                ans=n
                for i in range(1,n):
                    ans=max(ans,(i*f(n-i,dp)))
                dp[n]=ans
                return ans
        rans=0
        dp=[-1 for i in range(0,n+5)]
        for i in range(1,n):
            rans=max(rans,(i*f(n-i,dp)))
        return rans

File: 343-integer-break/343-integer-break.cpp
class Solution {
public:
    int dp[60];
    int f(int n)
    {
        if(n<=2) return 1;
        else if(dp[n]!=-1) return dp[n];
        else
        {
            int ans=0;
            for(int i=1;i<=n-1;i++)
            {
                ans=max({ans,i*(max(f(n-i),(n-i)))});
            }
            // cout<<n<<" "<<ans<<"\n";
            return dp[n]=ans;
        }
    }
    
    int integerBreak(int n) 
    {
        memset(dp,-1,sizeof(dp));
        return f(n);   
    }
};